### PR TITLE
Space Sensors Improvements

### PR DESCRIPTION
**Pull request by mkerensky**
+ Bugfix Corrects an issue where sensor detections in space battles create an infinite loop of entities, eventually consuming all available memory

+ Improvement Thermal/Optical sensors now properly establish firing solutions per SO rules

+ Bugfix Fighter squadrons no longer suffer a PSR penalty for "more than 2x safe thrust" in space battles

### CODE REVIEW

**Comment by HammerGS** [comment]
This pull request **introduces 4 alerts** when merging de0aa7c6aca50d3fe3978a366cf6ce047a9dd359 into 0c93c36d79c1995022ae3768629721e869503941 - [view on LGTM.com](https://lgtm.com/projects/g/MegaMek/megamek/rev/pr-cd9c5ea0da8412584327eb779086533f96291f28)

**new alerts:**

* 2 for Spurious Javadoc @param tags
* 2 for Dereferenced variable may be null

---

This project has automated code review enabled, but doesn't use the <a href="https://github.com/apps/lgtm-com">LGTM GitHub App</a>. Migrate over by <a href="https://github.com/apps/lgtm-com/installations/new">installing the app</a>. Read about the <a href="https://lgtm.com/blog/announcing_github_apps_migration">benefits of migrating to GitHub Apps</a> in the blog.

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by HammerGS** [comment]
This pull request **introduces 2 alerts** when merging 9c008535d5a4ac7d180e502b0d446d663be83505 into 0c93c36d79c1995022ae3768629721e869503941 - [view on LGTM.com](https://lgtm.com/projects/g/MegaMek/megamek/rev/pr-fb0b8eb33d063a204dc1246c8c273847eb667c2a)

**new alerts:**

* 2 for Dereferenced variable may be null

---

This project has automated code review enabled, but doesn't use the <a href="https://github.com/apps/lgtm-com">LGTM GitHub App</a>. Migrate over by <a href="https://github.com/apps/lgtm-com/installations/new">installing the app</a>. Read about the <a href="https://lgtm.com/blog/announcing_github_apps_migration">benefits of migrating to GitHub Apps</a> in the blog.

---

*Comment posted by [LGTM.com](https://lgtm.com)*

File: megamek/src/megamek/common/Compute.java [hunk 1]
```diff
@@ -4290,84 +4272,101 @@ private static int calcSensorShadow(IGame game, Entity ae,
...
                 || detector.getPosition() == null) {
-            detector.firingSolutions.removeAllElements();
+            detector.clearFiringSolutions();
             return;
         }
-        for (Entity target : detector.firingSolutions) {
-            //If the target is dead, has no position or has flown offboard, remove it
+        for (int id : detector.getFiringSolutions()) {
+            Entity target = game.getEntity(id);
+            //The target should be removed if it's off the board for any of these reasons
```
Comment (sixlettervariables) [inline megamek/src/megamek/common/Compute.java]: Like above, will target always be non-null?

File: megamek/src/megamek/common/Compute.java [hunk 2]
```diff
@@ -4290,84 +4272,101 @@ private static int calcSensorShadow(IGame game, Entity ae,
...
                 || detector.isOffBoard()
                 || detector.getPosition() == null) {
-            detector.firingSolutions.removeAllElements();
+            detector.clearSensorContacts();
             return;
         }
-        for (Entity target : detector.sensorContacts) {
-            //If the target is dead, has no position or has flown offboard, remove it
+        for (int id : detector.getSensorContacts()) {
+            Entity target = game.getEntity(id);
```
Comment (sixlettervariables) [inline megamek/src/megamek/common/Compute.java]: Same question regarding target.

File: megamek/src/megamek/common/Entity.java [hunk 3]
```diff
@@ -15565,5 +15567,88 @@ public int modifyPhysicalDamageForMeleeSpecialist() {
...
+     */
+    public void addSensorContact(int targetId) {
+        sensorContacts.add(targetId);
+    }
+    
+    /**
+     * Removes the specified target entity's ID from this entity's sensorContacts
+     * @param targetId the ID number of the target entity to remove
+     */
+    public void removeSensorContact(int targetId) {
```
Comment (sixlettervariables) [inline megamek/src/megamek/common/Entity.java]: Perhaps make this:
```java
public void removeSensorContacts(Collection<int> targetIds) {
    sensorContacts.removeAll(targetIds);
}
```
Then you can just pass `toRemove` directly to this, no need to check for `size >= 1`.

File: megamek/src/megamek/common/Compute.java (hunk 1 above)
Comment (mkerensky) [reply megamek/src/megamek/common/Compute.java]: Good catch. I had added a null guard to a couple of similar checks, but missed this one.

File: megamek/src/megamek/common/Compute.java (hunk 2 above)
Comment (mkerensky) [reply megamek/src/megamek/common/Compute.java]: Fixed.

File: megamek/src/megamek/common/Entity.java (hunk 3 above)
Comment (mkerensky) [reply megamek/src/megamek/common/Entity.java]: Done.  Didn't recognize that I could do it this way.

**Comment by mkerensky** [comment]
All change requests made.

**Comment by HammerGS** [comment]
This pull request **introduces 2 alerts** when merging fd1654b470441ba6f95fe6c4550fc50b8bd2cdfd into e5acb67405f4588931fbaeab1d1c022d41f823a3 - [view on LGTM.com](https://lgtm.com/projects/g/MegaMek/megamek/rev/pr-f9feda5ce505481327e7296b74bb1d41289561fd)

**new alerts:**

* 2 for Dereferenced variable may be null

---

This project has automated code review enabled, but doesn't use the <a href="https://github.com/apps/lgtm-com">LGTM GitHub App</a>. Migrate over by <a href="https://github.com/apps/lgtm-com/installations/new">installing the app</a>. Read about the <a href="https://lgtm.com/blog/announcing_github_apps_migration">benefits of migrating to GitHub Apps</a> in the blog.

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Pull request merged**